service/history/queues/executable.go

```diff
@@ -54,3 +66,250 @@ type (
+	// this determines if the executable should be retried within one submission to scheduler
+
+	// don't retry immediately for resource exhausted which may incur more load
+	// context deadline exceed may also suggested downstream is overloaded, so don't retry immediately
+	if common.IsResourceExhausted(err) || common.IsContextDeadlineExceededErr(err) {
+		return false
+	}
+
+	// ErrTaskRetry means mutable state is not ready for standby task processing
+	// there's no point for retrying the task immediately which will hold the worker corouinte
+	// TODO: change ErrTaskRetry to a better name
+	return err != consts.ErrTaskRetry
+}
+
+func (e *executableImpl) RetryPolicy() backoff.RetryPolicy {
+	// this is the retry policy for one submission
+	// not for calculating the backoff after the task is nacked
+	return schedulerRetryPolicy
+}
+
+func (e *executableImpl) Ack() {
+	e.Lock()
+	defer e.Unlock()
+
+	e.state = ctasks.TaskStateAcked
+
+	if e.shouldProcess {
+		e.scope.RecordDistribution(metrics.TaskAttemptTimer, e.attempt)
+		e.scope.RecordTimer(metrics.TaskLatency, time.Since(e.loadTime))
+		e.scope.RecordTimer(metrics.TaskQueueLatency, time.Since(e.GetVisibilityTime()))
+		e.scope.RecordTimer(metrics.TaskUserLatency, e.userLatency)
+		e.scope.RecordTimer(metrics.TaskNoUserLatency, time.Since(e.loadTime)-e.userLatency)
+		e.scope.RecordTimer(metrics.TaskNoUserQueueLatency, time.Since(e.GetVisibilityTime())-e.userLatency)
+	}
+}
+
+func (e *executableImpl) Nack(err error) {
+	submitted := false
+	attempt := e.Attempt()
+	if e.shouldResubmitOnNack(attempt, err) {
+		// we do not need to know if there any error during submission
+		// as long as it's not submitted, the execuable should be add
+		// to the rescheduler
+		submitted, _ = e.scheduler.TrySubmit(e)
+	}
+
+	if !submitted {
+		e.rescheduler.Add(e, e.backoffDuration(attempt))
+	}
+}
+
+func (e *executableImpl) Reschedule() {
+	e.Nack(nil)
+}
+
+func (e *executableImpl) State() ctasks.State {
+	e.Lock()
+	defer e.Unlock()
+
+	return e.state
+}
+
+func (e *executableImpl) GetPriority() int {
+	e.Lock()
+	defer e.Unlock()
+
+	return e.priority
+}
+
+func (e *executableImpl) SetPriority(priority int) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.priority = priority
+}
+
+func (e *executableImpl) Attempt() int {
+	e.Lock()
+	defer e.Unlock()
+
+	return e.attempt
+}
+
+func (e *executableImpl) Logger() log.Logger {
+	return e.logger
+}
+
+func (e *executableImpl) QueueType() QueueType {
+	return e.queueType
+}
+
+func (e *executableImpl) shouldResubmitOnNack(attempt int, err error) bool {
+	// this is an optimization for skipping rescheduler and retry the task sooner
+	// this can be useful for errors like unable to get workflow lock, which doesn't
+	// have to backoff for a long time and wait for the periodic rescheduling.
+	return e.IsRetryableError(err) && e.Attempt() < resubmitMaxAttempts
+}
+
+func (e *executableImpl) backoffDuration(attempt int) time.Duration {
```

Comment:
should be more specific -> rescheduleBackoff (to differentiate with retry within one submission. )